addons/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Comm.java

```diff
@@ -0,0 +1,1145 @@
+        System.arraycopy(bdata, 0, padded_data, 0, bdata.length);
+        System.arraycopy(padchars, 0, padded_data, bdata.length, padchars.length);
+        return padded_data;
+    }
+
+    /**
+     * This function does the GET http communication to the device
+     *
+     */
+    public byte[] getDataFromService(String service) {
+        byte[] responseBodyB64 = null;
+        int statusCode = 0;
+        // Create an instance of HttpClient.
+        synchronized (client) {
+            logger.debug("Starting receive connection...");
+
+            // Create a method instance.
+            GetMethod method = new GetMethod("http://" + device.getIP4Address() + service);
+
+            // Provide custom retry handler is necessary
+            method.getParams().setParameter(HttpMethodParams.RETRY_HANDLER,
+                    new DefaultHttpMethodRetryHandler(3, false));
+            // Set the right header
+            method.setRequestHeader("Accept", "application/json");
+            method.addRequestHeader("User-Agent", "TeleHeater/2.2.3");
+
+            try {
+                for (int i = 0; i < maxNbrRepeats.intValue() && statusCode != HttpStatus.SC_OK; i++) {
+                    // Execute the method.
+                    statusCode = client.executeMethod(method);
+                    // Check the status
+                    switch (statusCode) {
+                        case HttpStatus.SC_OK:
+                            break;
+                        case HttpStatus.SC_INTERNAL_SERVER_ERROR:
+                            /* Unknown problem with the device, wait and try again */
+                            logger.warn("HTTP GET failed: 500, internal server error, repeating.. ");
+                            Thread.sleep(100L * i + 1);
+                            continue;
+                        case HttpStatus.SC_FORBIDDEN:
+                            /* Service is available but not readable */
+                            byte[] test = new byte[1];
+                            return test;
+                        case HttpStatus.SC_NOT_FOUND:
+                            /* Should only happen on discovery */
+                            return null;
+                        default:
+                            logger.error("HTTP GET failed: {}", method.getStatusLine());
+                            return null;
+                    }
+                }
+                device.setCharSet(method.getResponseCharSet());
+                // Read the response body.
+                responseBodyB64 = ByteStreams.toByteArray(method.getResponseBodyAsStream());
+            } catch (HttpException e) {
+                logger.debug("Fatal protocol violation: {}", e.getMessage());
+            } catch (InterruptedException e) {
+                logger.debug("Sleep was interrupted: {}", e.getMessage());
+            } catch (IOException e) {
+                logger.debug("Fatal transport error: {}", e.getMessage());
```

Comment:
Should you update the state of something now it seems that it failed, should you maybe throw a more specific exception to the caller, because just passing null back is rude. This applies to all the catches here.

Reply:
In usual mode the user should see that there is something wrong. Sometimes the communication to the device fails but almost every time it repairs itself. So I would like to let it like is. The null return value is handled in the caller class.